src/t_zset.c

```diff
@@ -2184,6 +2200,177 @@ inline static void zunionInterAggregate(double *target, double val, int aggregat
+    }
+
+    dictReleaseIterator(di);
+
+    return maxelelen;
+}
+
+static void zdiffAlgorithm1(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    /* DIFF Algorithm 1:
+     *
+     * We perform the diff by iterating all the elements of the first set,
+     * and only adding it to the target set if the element does not exist
+     * into all the other sets.
+     *
+     * This way we perform at max N*M operations, where N is the size of
+     * the first set, and M the number of sets.
+     *
+     * There is also a O(K*log(K)) cost for adding the resulting elements
+     * to the target set, where K is the final size of the target set.
+     *
+     * The final complexity of this algorithm is O(N*M + K*log(K)). */
+    int j;
+    zsetopval zval;
+    zskiplistNode *znode;
+    sds tmp;
+
+    /* With algorithm 1 it is better to order the sets to subtract
+     * by decreasing size, so that we are more likely to find
+     * duplicated elements ASAP. */
+    qsort(src+1,setnum-1,sizeof(zsetopsrc),zuiCompareByRevCardinality);
+
+    memset(&zval, 0, sizeof(zval));
+    zuiInitIterator(&src[0]);
+    while (zuiNext(&src[0],&zval)) {
+        double score, value;
+        int exists = 0;
+
+        score = zval.score;
+        if (isnan(score)) score = 0;
+
+        for (j = 1; j < setnum; j++) {
+            /* It is not safe to access the zset we are
+             * iterating, so explicitly check for equal object.
+             * This check isn't really needed anymore since we already
+             * check for a duplicate set in the zsetChooseDiffAlgorithm
+             * function, but we're leaving it for future-proofing. */
+            if (src[j].subject == src[0].subject ||
+                zuiFind(&src[j],&zval,&value)) {
+                exists = 1;
+                break;
+            }
+        }
+
+        if (!exists) {
+            tmp = zuiNewSdsFromValue(&zval);
+            znode = zslInsert(dstzset->zsl,score,tmp);
+            dictAdd(dstzset->dict,tmp,&znode->score);
+            if (sdslen(tmp) > *maxelelen) *maxelelen = sdslen(tmp);
+        }
+    }
+}
+
+
+static void zdiffAlgorithm2(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    /* DIFF Algorithm 2:
+     *
+     * Add all the elements of the first set to the auxiliary set.
+     * Then remove all the elements of all the next sets from it.
+     *
+     * This is O(N + K*log(K)) where N is the sum of all the elements in every
+     * set, and K is the size of the resulting set.
+     *
+     * There is also a O(K) cost at the end for finding the largest element
+     * size, but this doesn't change the algorithm complexity. */
+    int j;
+    int cardinality = 0;
+    zsetopval zval;
+    zskiplistNode *znode;
+    sds tmp;
+
+    for (j = 0; j < setnum; j++) {
+        if (zuiLength(&src[j]) == 0) continue;
+
+        memset(&zval, 0, sizeof(zval));
+        zuiInitIterator(&src[j]);
+        while (zuiNext(&src[j],&zval)) {
```

Comment:
same, going backwards will make zslInsert faster

Reply:
Here I had an extra concern, which is the impact that this will have on `zsetslDel`. Should we iterate forward when deleting the elements? This would be simple, we could just move loop to inside the if that checks if it's the first set or not, iterating backwards in one case, and forward in the other.

Reply:
you mean `zslDelete`? looking at it, and considering we can't know the order of elements between efferent sets, i'm not sure if it has any principal difference.

maybe for the (possibly common case) that we have just two sets, and the majority of the elements are being deleted, it's best to iterate from head to tail. (constantly deleting the first element and no need for long searches).

maybe instead of changing the ZDIFF code, it's better to try to have zslInsert detect an insertion to the tail and make it efficient for that case too.
this way no one else will have to worry to scan insertions in the efficient direction.

Reply:
Seems like a better idea indeed. I'll try to look into it, but I'll have to better understand the skiplist data structure, because I read the zslInsert code and couldn't see an easy way to optimize for that case. The only idea I had was to add a "hint" so that it assumes the element is greater than all, but I don't think it's possible to just reverse the loops, since it would change the way the "rank" array in the function is calculated.

Reply:
It just ocurred to me: since this is an optimization that affects other commands, maybe it should be in a different pull request?

Reply:
yeah, it can (be in a different PR).
the reason to consider it now, is because we added a new inefficient code in ZDIFF.
but even if we'll end up concluding that zslInsert should remain as is, and decide to change the ZDIFF code to run backwards, we can do that in a different PR too.

p.s. the skiplist has a tail pointer, it should be easy to see that the new entry is greater than the tail. not sure about the efficiency of insertion yet.

anyway, if you rather lave that out for now i'm ok with it.
in which case we just need to merge my documentation changes (assuming you confirm that they're correct).

Reply:
As it's a deeper algorithmic change that affects other commands, it should fit well as a separate pull request, but I don't have anything against doing it here. I've finished everything else, so I think it's an option to consider. But anyway, I'll focus on the optimization of zslInsert from now on!

Reply:
let's make it in a different PR. this one is already going too long, and since the change isn't gonna affect the API, and might even be in a different are in the code with implications to other parts of redis...

Reply:
@felipou will you have time to try to handle that soon?
i wanna make ZDIFF efficient for 6.2 one way or another (we have about 2 weeks before an RC).

Reply:
I'll try to come up with something today, at least an initial PR that we can build on top of, that way if I don't have much time in the next weeks someone can pick up where I left. But I believe I can get something working by the next weekend.

Reply:
I just reviewed the insert algorithm for skip lists, and if I understand correctly, there is no way to efficiently insert backwards. To insert, we need to update the forward pointers in all levels, so we need to traverse the list anyway. The backwards pointer is only in the first level, and is only an optimization for ZREVRANGE to traverse the list backwards. If we created backward pointers for all levels, we would end up having to traverse the list both ways when inserting (and deleting and updating).

I'll try to come up with the backward equivalents of zuiInitIterator and zuiNext instead, so that we can modify the ZDIFF code directly.

Reply:
Just finished this PR: https://github.com/redis/redis/pull/8105